mautic/mautic
# Improved JS tracking code and fixed bug that would not allow multiple mt() calls

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Support for calling mt() multiple times only worked if the second mt() was wrapped in a setTimeout(). And then only for the first and second calls. Subsequent calls to mt() or back to back calls to mt() would fail. This PR fixes that.

It also now uses CORS by default to deliver the event. If the event fails, it falls back to creating a tracking pixel. This eliminates the need to make subsequent CORS requests to get the contact ID.

It also fixes a bug where the tracking pixel and/or CORS urls were forced to http if embedded into a http page but Mautic is under https.

It stores the contact ID and tracking pixel and appends it to the subsequent requests to prevent multiple visitor contacts from being created due to browsers not accepting cookies by default. 

It adds an option in Configuration -> Landing Page Settings to opt into prior behavior of tracking contacts by IP. 

Finally it prevents page hits and contacts from being created for requests from "do not track IPs." 

#### Steps to test this PR:
1. Embed the tracking code into a web page and add the call to mt() several times. Look at the contact's page hits. Each call to mt() should be registered. 
2. Configure CORS and do the same. A call to /mtc/event should be made instead of mtracking.gif. Again, each hit should be registered. 
3. Access Mautic on https and embed the tracking code on http. Tracking should still work.
4. Configure the browser to only accepts cookies from sites previously visited. Clear the browser's cache (cookies, etc) if used to access Mautic before. Embed the tracking code on a domain other than that of Mautic (be sure to use dev if on local so that localhost IPs are acknowledged). Visit the page and then look at anonymous contacts in Mautic. Only one anonymous should be created.
5. Enable the new track by IP option. Visit a tracked page. Clear all the cookies and local session and hit it again. The same contact should be used for subsequent requests. Disable the option and repeat (be sure to clear browser cache to avoid the new localStorage fix). This time a new contact should have been created.
6. Add your IP to Configuration -> System Settings -> Do Not Track IPs (if on localhost, use 127.0.0.1). Clear browser cache and visit a tracked page. The hit nor contact should be created (good to avoid spam IP or known bot IPs). 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Repeat #1 above but only the first mt() will be recorded. 
2. Repeat #3 above and will get a ` Response for preflight is invalid` error in the browsers dev console
